### PR TITLE
fix(noq): do a best-effort transmit on connection close

### PR DESCRIPTION
**Pull request by colinmarc**
## Description

If we're shutting down ungracefully, we can make a best-effort attempt to write the CONNECTION_CLOSE to the socket and notify the peer. If we're in the middle of graceful shutdown, it doesn't hurt, either.

This can be taken advantage of by callers to do a quicker shutdown if they know the drain period isn't necessary. Discussion here: https://github.com/n0-computer/iroh/issues/4201

### CODE REVIEW

**Review by matheus23 — COMMENTED** [review]
Thank you for looking into this @colinmarc :)

Before we put in more work, I need to check back with @flub whether our ideas in https://github.com/n0-computer/iroh/issues/4201 are sound & desirable.

Other than that:
- This needs a test! Ideally one that sets up a `Runtime` similar to [the one we're using in iroh](https://github.com/n0-computer/iroh/blob/main/iroh/src/runtime.rs): Most importantly one that uses a `TaskTracker` and fairly aggressively closes all tasks (in this case those will be the connection and endpoint drivers).
- Rather than driving the transmit in `close`, I'd like to drive the transmit only when the `ConnectionDriver` is dropped. Otherwise I worry the `drive_transmit` from the connection driver and the `close` call can get in each others way. I think there's an invariant that the connection driver is the only thing that calls `drive_transmit` and it never misses the consecutive calls to that function. The right place for calling `drive_transmit` might be in the `impl Drop for ConnectionRef` right after the `conn.implicit_close(..)`.

**Comment by colinmarc** [comment]
> Before we put in more work, I need to check back with @flub whether our ideas in https://github.com/n0-computer/iroh/issues/4201 are sound & desirable.

No worries, I just wanted to have something concrete to discuss :) This is only half of a workaround for that issue anyway.

> The right place for calling drive_transmit might be in the impl Drop for ConnectionRef right after the conn.implicit_close(..).

That makes sense! Moved it to the drop for `ConnectionRef`. I wasn't sure if you meant inside the conditional, but that would mean we don't get this optimization for an *explicit* close/drop.

File: noq/src/connection.rs [hunk 1]
```diff
@@ -1233,6 +1233,14 @@ impl Drop for ConnectionRef {
...
             conn.implicit_close(&self.shared);
         }
+
+        // We also flush any pending packets, e.g. the CONNECTION_CLOSE, if we're able to do so
+        // synchronously. This is best-effort; we won't wait for them to be acked, but we have a
+        // decent chance of notifying the peer in the case that we're currently shutting down
+        // ungracefully.
+        let waker = std::task::Waker::noop();
+        let mut cx = std::task::Context::from_waker(waker);
+        let _ = conn.drive_transmit(&mut cx);
```
Comment (flub) [inline noq/src/connection.rs:1243]: It is kind of tempting to do this. But also, there's very little guarantee. E.g. even if a CONNECTION_CLOSE was scheduled for transmit, it is supposed to respect congestion control and might well not be sent (that could be buggy currently, I recall spotting some CONNECTION_CLOSE packets that did not respect congestion control).

I think at the end of the day when ungraceful things happen the remote has to deal with things just timing out. I think the edge-cases for which this would help are not really worth this extra complexity. I'd rather people discover these problems faster in tests if their code is genuinely misbehaving.

So I'm tempted to close this without merging.

Comment (matheus23) [reply noq/src/connection.rs:1243]: Yeah - I also ran some tests to see if this would actually help.
The idea was to make graceful close work when you just `drop(endpoint)`, but keep the process alive a bit.

I didn't dive into the details of why it didn't quite work, but I didn't get the expected result. At least not when the `AsyncUdpSocket` implementation is the one it is in iroh (I haven't tested it with a normal `UdpSocket` in noq).

So yeah - I think this might need a bit more thorough work and looking into how this would actually work.

W.r.t. congestion control I think it is completely fine to not send the CONNECTION_CLOSE if CC doesn't allow it. I expect that in most cases CC will actually allow the close frame, though I admit relying on that behavior is a bad idea.

Comment (colinmarc) [reply noq/src/connection.rs:1243]: To be clear, I wouldn't really expect it to work with anything *other* than a normal udp socket. One thing I'm not clear on - with iroh, after migration off the relay, isn't that the case?

Comment (flub) [reply noq/src/connection.rs:1243]: yes, once iroh selected the direct path it should behave the same as noq with a normal udp socket as far as this PR is concerned.
